File: coordinate.py
import numpy as np
from numpy.linalg import norm
from math import sin,cos,tan,asin,acos,atan,atan2,degrees,radians,sqrt
from numba import jit
import pymap3d as pm
from utils import *

@jit('f8[:](f8[:],f8[:])',nopython=True)
def quatmult(q, p):
    qp0 = q[0]*p[0] - q[1]*p[1] - q[2]*p[2] - q[3]*p[3]
    qp1 = q[1]*p[0] + q[0]*p[1] - q[3]*p[2] + q[2]*p[3]
    qp2 = q[2]*p[0] + q[3]*p[1] + q[0]*p[2] - q[1]*p[3]
    qp3 = q[3]*p[0] - q[2]*p[1] + q[1]*p[2] + q[0]*p[3]
    return np.array([qp0, qp1, qp2, qp3])

@jit('f8[:](f8[:])',nopython=True)
def conj(q):
    return np.array([q[0], -q[1], -q[2], -q[3]])

@jit('f8[:](f8[:])',nopython=True)
def normalize(v):
    return v / norm(v)


@jit('f8[:](f8[:],f8[:])',nopython=True)
def quatrot(q, v):
    vq  = np.array((0.0, v[0], v[1], v[2]))
    rvq = quatmult(conj(q), quatmult(vq, q))
    return rvq[1:4]

@jit(nopython=True)
def dcm_from_quat(q):
    C = np.zeros((3,3))
    C[0,0] = q[0]**2+q[1]**2-q[2]**2-q[3]**2
    C[0,1] = 2.0 * (q[1]*q[2] + q[0]*q[3])
    C[0,2] = 2.0 * (q[1]*q[3] - q[0]*q[2])

    C[1,0] = 2.0 * (q[1]*q[2] - q[0]*q[3])
    C[1,1] = q[0]**2-q[1]**2+q[2]**2-q[3]**2
    C[1,2] = 2.0 * (q[2]*q[3] + q[0]*q[1])

    C[2,0] = 2.0 * (q[1]*q[3] + q[0]*q[2])
    C[2,1] = 2.0 * (q[2]*q[3] - q[0]*q[1])
    C[2,2] = q[0]**2-q[1]**2-q[2]**2+q[3]**2

    return C

@jit(nopython=True)
def quat_from_dcm(C):
    if (1.0 + C[0,0] + C[1,1] + C[2,2]) < 0.0:
        print("quaternion conversion error")
        return np.array((1.0, 0.0, 0.0, 0.0))

    q0 = 0.5 * sqrt(1.0 + C[0,0] + C[1,1] + C[2,2])
    q1 = 0.25 / q0 * (C[1,2] - C[2,1])
    q2 = 0.25 / q0 * (C[2,0] - C[0,2])
    q3 = 0.25 / q0 * (C[0,1] - C[1,0])

    return np.array((q0, q1, q2, q3))   

@jit('f8[:](f8[:],f8)',nopython=True)
def ecef2eci(xyz_in, t):
    omega_earth_rps = 7.2921151467e-5
    xyz_out = np.zeros(3)
    xyz_out[0] = xyz_in[0] * cos(omega_earth_rps * t) - xyz_in[1] * sin(omega_earth_rps * t)
    xyz_out[1] = xyz_in[0] * sin(omega_earth_rps * t) + xyz_in[1] * cos(omega_earth_rps * t)
    xyz_out[2] = xyz_in[2]
    return xyz_out


@jit('f8[:](f8[:],f8)',nopython=True)
def eci2ecef(xyz_in, t):
    omega_earth_rps = 7.2921151467e-5
    xyz_out = np.zeros(3)
    xyz_out[0] = xyz_in[0] * cos(omega_earth_rps * t) + xyz_in[1] * sin(omega_earth_rps * t)
    xyz_out[1] =-xyz_in[0] * sin(omega_earth_rps * t) + xyz_in[1] * cos(omega_earth_rps * t)
    xyz_out[2] = xyz_in[2]
    return xyz_out


@jit('f8[:](f8[:],f8[:],f8)',nopython=True)
def vel_ecef2eci(vel_in, pos_in, t):
    
    omega_earth_rps = 7.2921151467e-5
    pos_eci = ecef2eci(pos_in, t)
    vel_ground_eci = ecef2eci(vel_in, t)
    
    vel_rotation_eci = np.cross(np.array([0, 0, omega_earth_rps]), pos_eci)
    
    return vel_ground_eci + vel_rotation_eci


@jit('f8[:](f8[:],f8[:],f8)',nopython=True)
def vel_eci2ecef(vel_in, pos_in, t):
    
    omega_earth_rps = 7.2921151467e-5
    
    vel_rotation_eci = np.cross(np.array([0, 0, omega_earth_rps]), pos_in)
    vel_ground_eci = vel_in - vel_rotation_eci
    
    return eci2ecef(vel_ground_eci, t)

@jit('f8[:](f8)',nopython=True)
def quat_eci2ecef(t):
    omega_earth_rps = 7.2921151467e-5
    return np.array([cos(omega_earth_rps*t/2.0), 0.0, 0.0, sin(omega_earth_rps*t/2.0)])

@jit('f8[:](f8)',nopython=True)
def quat_ecef2eci(t):
    return conj(quat_eci2ecef(t))

@jit('f8[:](f8[:])',nopython=True)
def quat_ecef2nedc(pos_ecef):
    l = atan2(pos_ecef[1], pos_ecef[0])
    p = asin(pos_ecef[2] / norm(pos_ecef))
    c_hl = cos(l/2.0)
    s_hl = sin(l/2.0)
    c_hp = cos(p/2.0)
    s_hp = sin(p/2.0)
    
    quat_ecef2ned = np.zeros(4)
    quat_ecef2ned[0] = c_hl * (c_hp - s_hp) / sqrt(2.0)
    quat_ecef2ned[1] = s_hl * (c_hp + s_hp) / sqrt(2.0)
    quat_ecef2ned[2] =-c_hl * (c_hp + s_hp) / sqrt(2.0)
    quat_ecef2ned[3] = s_hl * (c_hp - s_hp) / sqrt(2.0)
    
    return quat_ecef2ned


@jit('f8[:](f8[:])',nopython=True)
def quat_ecef2nedg(pos_ecef):
    
    p,l,_ = ecef2geodetic(pos_ecef[0], pos_ecef[1], pos_ecef[2])
    p = radians(p)
    l = radians(l)
    
    c_hl = cos(l/2.0)
    s_hl = sin(l/2.0)
    c_hp = cos(p/2.0)
    s_hp = sin(p/2.0)
    
    quat_ecef2ned = np.zeros(4)
    quat_ecef2ned[0] = c_hl * (c_hp - s_hp) / sqrt(2.0)
    quat_ecef2ned[1] = s_hl * (c_hp + s_hp) / sqrt(2.0)
    quat_ecef2ned[2] =-c_hl * (c_hp + s_hp) / sqrt(2.0)
    quat_ecef2ned[3] = s_hl * (c_hp - s_hp) / sqrt(2.0)
    
    return quat_ecef2ned

@jit('f8[:](f8[:])',nopython=True)
def quat_nedg2ecef(pos_ecef):
    return conj(quat_ecef2nedg(pos_ecef))

@jit('f8[:](f8[:])',nopython=True)
def quat_nedc2ecef(pos_ecef):
    return conj(quat_ecef2nedc(pos_ecef))


@jit('f8[:](f8[:],f8)',nopython=True)
def quat_eci2nedg(pos, t):
    return quatmult(quat_eci2ecef(t), quat_ecef2nedg(eci2ecef(pos,t)))

@jit('f8[:](f8[:],f8)',nopython=True)
def quat_eci2nedc(pos, t):
    return quatmult(quat_eci2ecef(t), quat_ecef2nedc(eci2ecef(pos,t)))


@jit('f8[:](f8[:],f8)',nopython=True)
def quat_nedg2eci(pos, t):
    return conj(quat_eci2nedg(pos, t))

@jit('f8[:](f8[:],f8)',nopython=True)
def quat_nedc2eci(pos, t):
    return conj(quat_eci2nedc(pos, t))


@jit('f8[:](f8[:],f8[:])',nopython=True)
def quat_eci2guide(pos, vel):
    
    z_dir = normalize(-pos)
    x_dir = normalize(np.cross(vel, pos))
    y_dir = np.cross(z_dir,x_dir)
    
    q0 = 0.5 * sqrt(1.0 + x_dir[0] + y_dir[1] + z_dir[2])
    q1 = 0.25 / q0 * (y_dir[2] - z_dir[1])
    q2 = 0.25 / q0 * (z_dir[0] - x_dir[2])
    q3 = 0.25 / q0 * (x_dir[1] - y_dir[0])
    
    return normalize(np.array((q0, q1, q2, q3)))


@jit('f8[:](f8[:],f8[:],f8[:])',nopython=True)
def quat_guide2body(pos, vel, quat_eci2body):
    return quatmult(conj(quat_eci2guide(pos, vel)), quat_eci2body)



@jit('f8[:](f8,f8,f8)',nopython=True)
def quat_from_euler(az, el, ro):
    qz = np.array([cos(radians(az/2)), 0.0, 0.0, sin(radians(az/2))])
    qy = np.array([cos(radians(el/2)), 0.0, sin(radians(el/2)), 0.0])
    qx = np.array([cos(radians(ro/2)), sin(radians(ro/2)), 0.0, 0.0])

    return quatmult(qz, quatmult(qy, qx))

@jit(nopython=True)
def gravity(pos):
    x,y,z = pos
    
    a = 6378137
    f = 1.0 / 298.257223563
    mu = 3.986004418e14
    J2 = 1.082628e-3
    
    r = norm(pos)
    p2 = x**2+y**2
    
    fx = mu * (-x / r**3 + J2 * a**2 * x / r**7 * (6.0 * z**2 - 1.5 * p2))
    fy = mu * (-y / r**3 + J2 * a**2 * y / r**7 * (6.0 * z**2 - 1.5 * p2))
    fz = mu * (-z / r**3 + J2 * a**2 * z / r**7 * (3.0 * z**2 - 4.5 * p2))

    return np.array([fx, fy, fz])

@jit('f8[:](f8[:],f8[:],f8)',nopython=True)
def quat_nedg2body(quat_eci2body, pos, t):
    q = quat_eci2nedg(pos, t)
    return quatmult(conj(q), quat_eci2body)

@jit('f8[:](f8[:])',nopython=True)
def euler_from_quat(q):
    if 2.0*(q[0]*q[2]-q[3]*q[1]) >= 1.0:
        el = np.pi/2
        az = 0.0
        ro = 0.0
    else:
        az = atan2(2.0 * (q[0]*q[3]+q[1]*q[2]), 1.0-2.0*(q[2]**2+q[3]**2))
        el =  asin(2.0 * (q[0]*q[2]-q[3]*q[1]))
        ro = atan2(2.0 * (q[0]*q[1]+q[2]*q[3]), 1.0-2.0*(q[1]**2+q[2]**2))
    return np.rad2deg(np.array([az, el, ro]))

@jit(nopython=True)
def euler_from_dcm(C):
    el = asin(-C[0,2])
    if cos(el) < 0.0001:
        az = 0.0
        ro = 0.0
    else:
        az = atan2(C[0,1], C[0,0])
        ro = atan2(C[1,2], C[2,2])
    return np.rad2deg(np.array([az, el, ro]))

@jit(nopython=True)
def dcm_from_thrustvector(pos_eci, u):

    xb_dir = normalize(u)
    if u.dot(normalize(pos_eci)) > 0.999999:
        yb_dir = normalize(np.cross(np.array([0.0, 0.0, 1.0]), u))
    else:
        yb_dir = normalize(np.cross(u, normalize(pos_eci)))
    zb_dir = np.cross(u, yb_dir)

    return np.vstack((xb_dir, yb_dir, zb_dir))  

def eci2geodetic(pos_in, t):
    pos_ecef = eci2ecef(pos_in, t)
    return pm.ecef2geodetic(pos_ecef[0],pos_ecef[1],pos_ecef[2])


@jit('f8[:](f8[:],f8[:])', nopython=True)
def orbital_elements(r_eci, v_eci):
    GMe = 3.986004418e14
    
    nr = normalize(r_eci)
    v_r = nr.dot(v_eci)
    v_n = norm(v_eci - v_r * normalize(nr))
    
    c_eci = np.cross(r_eci, v_eci) # orbit plane vector
    f_eci = np.cross(v_eci, c_eci) - GMe * normalize(r_eci) # Laplace vector
    
    c1_eci = normalize(c_eci)
    
    inclination_rad = acos(c1_eci.dot(np.array([0.0,0.0,1.0])))
    
    if inclination_rad > 1e-10:
        ascending_node_rad = atan2(c1_eci[0], -c1_eci[1])
        n_eci = np.array([cos(ascending_node_rad), sin(ascending_node_rad), 0.0]) # direction of ascending node
        argument_perigee = acos(n_eci.dot(normalize(f_eci)))
        if f_eci[2] < 0:
            argument_perigee *= -1.0
    else:
        ascending_node_rad = 0.0
        argument_perigee = atan2(f_eci[1], f_eci[0])
    
    p = norm(c_eci)**2 / GMe # semi-latus rectum
    e = norm(f_eci) / GMe # eccentricity
    a = p / (1.0 - e**2) # semi-major axis
    
    true_anomaly_rad = acos(normalize(f_eci).dot(normalize(r_eci)))
    if v_eci.dot(r_eci) < 0.0:
        true_anomaly_rad = 2.0*np.pi - true_anomaly_rad
        
    return np.array([a, e, degrees(inclination_rad), degrees(ascending_node_rad), degrees(argument_perigee), degrees(true_anomaly_rad)])

@jit(nopython=True)
def posvel_from_orbital_elements(elem):
    
    a = elem[0]
    e = elem[1]
    inclination_rad = radians(elem[2])
    ascending_node_rad = radians(elem[3])
    argument_perigee_rad = radians(elem[4])
    true_anomaly_rad = radians(elem[5])
    
    

File: src/Coordinate.cpp
// Copyright (c) 2022 Interstellar Technologies Inc.
// All rights reserved.

#include "Coordinate.hpp"

#include <cmath>
#include <iostream>

using Eigen::Matrix;
using Eigen::Matrix3d;
using Eigen::Quaterniond;
using Eigen::Vector3d;
using std::acos;
using std::atan2;
using std::cos;
using std::sin;
using std::sqrt;

Vector3d Coordinate::ecef2eci(Vector3d xyz_ecef, double t) {
  Vector3d out;
  out(0) = xyz_ecef(0) * cos(Earth::omega_earth_rps * t) -
           xyz_ecef(1) * sin(Earth::omega_earth_rps * t);
  out(1) = xyz_ecef(0) * sin(Earth::omega_earth_rps * t) +
           xyz_ecef(1) * cos(Earth::omega_earth_rps * t);
  out(2) = xyz_ecef(2);
  return out;
}

Vector3d Coordinate::eci2ecef(Vector3d xyz_eci, double t) {
  Vector3d out;
  out(0) = xyz_eci(0) * cos(Earth::omega_earth_rps * t) +
           xyz_eci(1) * sin(Earth::omega_earth_rps * t);
  out(1) = -xyz_eci(0) * sin(Earth::omega_earth_rps * t) +
           xyz_eci(1) * cos(Earth::omega_earth_rps * t);
  out(2) = xyz_eci(2);
  return out;
}

Vector3d Coordinate::vel_ecef2eci(Vector3d vel_ecef, Vector3d pos_ecef,
                                  double t) {
  Vector3d pos_eci = ecef2eci(pos_ecef, t);
  Vector3d vel_ground_eci = ecef2eci(vel_ecef, t);
  Vector3d omega_vec(0.0, 0.0, Earth::omega_earth_rps);
  return vel_ground_eci + omega_vec.cross(pos_eci);
}

Vector3d Coordinate::vel_eci2ecef(Vector3d vel_eci, Vector3d pos_eci,
                                  double t) {
  Vector3d omega_vec(0.0, 0.0, Earth::omega_earth_rps);
  return eci2ecef(vel_eci - omega_vec.cross(pos_eci), t);
}

Quaterniond Coordinate::quat_eci2ecef(double t) {
  Quaterniond out(cos(Earth::omega_earth_rps * t / 2.0), 0.0, 0.0,
                  sin(Earth::omega_earth_rps * t / 2.0));
  return out;
}

Quaterniond Coordinate::quat_ecef2eci(double t) {
  return quat_eci2ecef(t).conjugate();
}

Quaterniond Coordinate::quat_ecef2ned(Vector3d pos_ecef) {
  Vector3d geodetic = Earth::ecef2geodetic(pos_ecef);
  double c_hl = cos(geodetic(1) / 2.0);
  double s_hl = sin(geodetic(1) / 2.0);
  double c_hp = cos(geodetic(0) / 2.0);
  double s_hp = sin(geodetic(0) / 2.0);

  double q0 = c_hl * (c_hp - s_hp) / sqrt(2.0);
  double q1 = s_hl * (c_hp + s_hp) / sqrt(2.0);
  double q2 = -c_hl * (c_hp + s_hp) / sqrt(2.0);
  double q3 = s_hl * (c_hp - s_hp) / sqrt(2.0);
  Quaterniond q(q0, q1, q2, q3);
  return q;
}

Quaterniond Coordinate::quat_ned2ecef(Vector3d pos_ecef) {
  return quat_ecef2ned(pos_ecef).conjugate();
}

Quaterniond Coordinate::quat_eci2ned(Vector3d pos_eci, double t) {
  return quat_eci2ecef(t) * quat_ecef2ned(eci2ecef(pos_eci, t));
}

Quaterniond Coordinate::quat_ned2eci(Vector3d pos_eci, double t) {
  return quat_eci2ned(pos_eci, t).conjugate();
}

Quaterniond Coordinate::quat_ned2body(Quaterniond quat_eci2body,
                                      Vector3d pos_eci, double t) {
  return quat_ned2eci(pos_eci, t).conjugate() * quat_eci2body;
}

Quaterniond Coordinate::quat_from_euler_deg(double az_deg, double el_deg,
                                            double ro_deg) {
  double az = az_deg * M_PI / 180.0;
  double el = el_deg * M_PI / 180.0;
  double ro = ro_deg * M_PI / 180.0;
  Quaterniond q = Eigen::AngleAxisd(az, Vector3d::UnitZ()) *
                  Eigen::AngleAxisd(el, Vector3d::UnitY()) *
                  Eigen::AngleAxisd(ro, Vector3d::UnitX());
  return q;
}

Vector3d Coordinate::euler_from_quat(Quaterniond q) {
  Vector3d out = q.toRotationMatrix().eulerAngles(2, 1, 0);
  // EigenのeulerAnglesは値域が[0:pi]x[-pi:pi]x[-pi:pi]なので修正し、
  // [0,2pi]x[-pi/2,pi/2]x[-pi,pi]の範囲に収まるようにする
  if (std::fabs(out[1]) > M_PI / 2.0) {
    if (out[1] > 0.0) {
      out[1] = M_PI - out[1];
    } else {
      out[1] = -M_PI - out[1];
    }
    out[0] = M_PI + out[0];
    out[2] = M_PI + out[2];
  }
  out[0] = std::fmod(out[0], 2.0 * M_PI);
  if (out[0] < 0.0)
    out[0] += 2.0 * M_PI;
  out[2] = std::fmod(out[2] + M_PI, 2.0 * M_PI) - M_PI;
  return out;
}

Vector3d Coordinate::euler_from_dcm(Matrix3d C) {
  Vector3d out = C.transpose().eulerAngles(2, 1, 0);
  // EigenのeulerAnglesは値域が[0:pi]x[-pi:pi]x[-pi:pi]なので修正し、
  // [0,2pi]x[-pi/2,pi/2]x[-pi,pi]の範囲に収まるようにする
  if (std::fabs(out[1]) > M_PI / 2.0) {
    if (out[1] > 0.0) {
      out[1] = M_PI - out[1];
    } else {
      out[1] = -M_PI - out[1];
    }
    out[0] = M_PI + out[0];
    out[2] = M_PI + out[2];
  }
  out[0] = std::fmod(out[0], 2.0 * M_PI);
  if (out[0] < 0.0)
    out[0] += 2.0 * M_PI;
  out[2] = std::fmod(out[2] + M_PI, 2.0 * M_PI) - M_PI;
  return out;
}

Matrix3d Coordinate::dcm_from_quat(Quaterniond q) {
  Matrix3d C = q.toRotationMatrix().transpose();
  return C;
}

Quaterniond Coordinate::quat_from_dcm(Matrix3d C) {
  Quaterniond q(C.transpose());
  return q;
}

Matrix3d Coordinate::dcm_from_thrustvector(Vector3d thrustvec_eci,
                                           Vector3d pos_eci) {
  Vector3d xb_dir = thrustvec_eci.normalized();
  Vector3d yb_dir;
  if (1.0 - xb_dir.dot(pos_eci.normalized()) < 1.0e-10) {
    Vector3d z_eci(0.0, 0.0, 1.0);
    yb_dir = z_eci.cross(xb_dir).normalized();
  } else {
    yb_dir = xb_dir.cross(pos_eci.normalized()).normalized();
  }
  Vector3d zb_dir = xb_dir.cross(yb_dir);
  Matrix3d out;
  out << xb_dir, yb_dir, zb_dir;  // 列ベクトルを行方向に並べる
  return out.transpose();
}

Quaterniond Coordinate::quat_from_thrustvector(Vector3d thrustvec_eci,
                                               Vector3d pos_eci) {
  return quat_from_dcm(dcm_from_thrustvector(thrustvec_eci, pos_eci));
}

Matrix<double, 6, 1> Coordinate::orbital_elements(Vector3d pos_eci,
                                                  Vector3d vel_eci) {
  Vector3d nr = pos_eci.normalized();

  Vector3d c_eci = pos_eci.cross(vel_eci);                 // orbit plane vector
  Vector3d f_eci = vel_eci.cross(c_eci) - Earth::mu * nr;  // Laplace vector

  Vector3d c1_eci = c_eci.normalized();
  Vector3d f1_eci = f_eci.normalized();

  double inclination_rad = acos(c1_eci(2));
  double ascending_node_rad = 0.0;
  double argument_perigee_rad = 0.0;

  if (inclination_rad > 1.0e-10) {
    ascending_node_rad = atan2(c1_eci(0), -c1_eci(1));
    Vector3d n_eci(cos(ascending_node_rad), sin(ascending_node_rad),
                   0.0);  // direction of ascending node
    argument_perigee_rad = acos(n_eci.dot(f1_eci));
    if (f_eci(2) < 0.0) {
      argument_perigee_rad *= -1.0;
    }
  } else {
    ascending_node_rad = 0.0;
    if (f_eci.norm() > 1.0e-10) {
      argument_perigee_rad = atan2(f_eci(1), f_eci(0));
    } else {
      argument_perigee_rad = 0.0;
    }
  }

  double p = c_eci.dot(c_eci) / Earth::mu;
  double e = f_eci.norm() / Earth::mu;
  double a = p / (1.0 - e * e);

  double true_anomaly_rad = acos(f1_eci.dot(nr));
  if (vel_eci.dot(pos_eci) < 0.0) {
    true_anomaly_rad = 2.0 * M_PI - true_anomaly_rad;
  }

  if (ascending_node_rad < 0.0)
    ascending_node_rad += 2.0 * M_PI;
  if (argument_perigee_rad < 0.0)
    argument_perigee_rad += 2.0 * M_PI;
  if (true_anomaly_rad < 0.0)
    true_anomaly_rad += 2.0 * M_PI;

  Matrix<double, 6, 1> out;
  out << a, e, inclination_rad, ascending_node_rad, argument_perigee_rad,
      true_anomaly_rad;
  return out;
}

Vector3d Coordinate::pos_from_orbital_elements(Matrix<double, 6, 1> elem) {
  double a = elem[0];
  double e = elem[1];
  double i = elem[2];
  double O = elem[3];
  double w = elem[4];
  double theta = elem[5];

  double p = a * (1.0 - e * e);
  Vector3d c1_eci(sin(i) * sin(O), -sin(i) * cos(O), cos(i));
  Vector3d f1_eci(cos(O) * cos(w) - sin(O) * cos(i) * sin(w),
                  sin(O) * cos(w) + cos(O) * cos(i) * sin(w), sin(i) * sin(w));
  Vector3d y1_eci = c1_eci.cross(f1_eci);

  return p / (1.0 + e * cos(theta)) *
         (cos(theta) * f1_eci + sin(theta) * y1_eci);
}

Vector3d Coordinate::vel_from_orbital_elements(Matrix<double, 6, 1> elem) {
  double a = elem[0];
  double e = elem[1];
  double i = elem[2];
  double O = elem[3];
  double w = elem[4];
  double theta = elem[5];

  double p = a * (1.0 - e * e);
  Vector3d c1_eci(sin(i) * sin(O), -sin(i) * cos(O), cos(i));
  Vector3d f1_eci(cos(O) * cos(w) - sin(O) * cos(i) * sin(w),
                  sin(O) * cos(w) + cos(O) * cos(i) * sin(w), sin(i) * sin(w));
  Vector3d y1_eci = c1_eci.cross(f1_eci);

  return sqrt(Earth::mu / p) *
         (-sin(theta) * f1_eci + (e + cos(theta)) * y1_eci);
}
